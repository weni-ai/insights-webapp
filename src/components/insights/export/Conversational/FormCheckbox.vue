<template>
  <ExportCheckboxs
    data-testid="conversational-form-checkbox"
    :modelFields="translatedModelFields"
    :selectedFields="translatedSelectedFields"
    :enabledModels="translatedEnabledModels"
    :modelFilters="modelFilters"
    :isLoading="isLoading"
    @model-toggle="handleModelToggle"
    @field-toggle="handleFieldToggle"
  />
</template>

<script setup lang="ts">
import ExportCheckboxs from '../ExportCheckboxs.vue';
import { useConversationalExport } from '@/store/modules/export/conversational/export';
import { storeToRefs } from 'pinia';
import { onMounted, ref, computed } from 'vue';

const conversationalExport = useConversationalExport();
const {
  updateModelFieldSelection,
  toggleModelEnabled,
  initializeDefaultFields,
} = conversationalExport;
const { model_fields, selected_fields, enabled_models, custom_widgets } =
  storeToRefs(conversationalExport);
const isLoading = ref(false);

const modelFilters = computed(() => []);

const getUniqueDisplayName = (widget: any): string => {
  const baseName = widget.name;
  const shortUuid = widget.uuid.slice(0, 8);

  const duplicates = custom_widgets.value.filter((w) => w.name === baseName);

  if (duplicates.length > 1) {
    return `${baseName} (${shortUuid})`;
  }

  return baseName;
};

const translatedSelectedFields = computed(() => {
  const fields = { ...selected_fields.value };

  if (custom_widgets.value.length > 0) {
    custom_widgets.value.forEach((widget) => {
      if (fields[widget.uuid]) {
        const fieldData = fields[widget.uuid];
        delete fields[widget.uuid];
        const uniqueName = getUniqueDisplayName(widget);
        fields[uniqueName] = fieldData;
      }
    });
  }

  return fields;
});

const translatedEnabledModels = computed(() => {
  return enabled_models.value.map((modelKey) => {
    const widget = custom_widgets.value.find((w) => w.uuid === modelKey);
    return widget ? getUniqueDisplayName(widget) : modelKey;
  });
});

const translatedModelFields = computed(() => {
  const fields = { ...model_fields.value };

  if (custom_widgets.value.length > 0) {
    custom_widgets.value.forEach((widget) => {
      if (fields[widget.uuid]) {
        const fieldData = fields[widget.uuid];
        delete fields[widget.uuid];
        const uniqueName = getUniqueDisplayName(widget);
        fields[uniqueName] = fieldData;
      }
    });
  }

  return fields;
});

const initializeFields = () => {
  isLoading.value = true;
  try {
    initializeDefaultFields();
  } catch (error) {
    console.error('Error initializing conversational fields', error);
  } finally {
    isLoading.value = false;
  }
};

const getModelKeyForStore = (displayName: string): string => {
  let widget = custom_widgets.value.find((w) => w.name === displayName);

  if (!widget) {
    const match = displayName.match(/^(.+) \(([a-f0-9-]{8})\)$/);
    if (match) {
      const [, baseName, shortUuid] = match;
      widget = custom_widgets.value.find(
        (w) => w.name === baseName && w.uuid.startsWith(shortUuid),
      );
    }
  }

  return widget ? widget.uuid : displayName;
};

const handleModelToggle = (modelName: string, enabled: boolean) => {
  const storeKey = getModelKeyForStore(modelName);
  toggleModelEnabled(storeKey, enabled);
};

const handleFieldToggle = (
  modelName: string,
  fieldName: string,
  selected: boolean,
) => {
  const storeKey = getModelKeyForStore(modelName);
  updateModelFieldSelection(storeKey, fieldName, selected);
};

onMounted(() => {
  initializeFields();
});
</script>
