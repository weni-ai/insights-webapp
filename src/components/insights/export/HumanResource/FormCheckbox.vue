<template>
  <ExportCheckboxs
    :modelFields="model_fields"
    :selectedFields="selected_fields"
    :enabledModels="enabled_models"
    :sectors="sectors"
    :queues="queues"
    :agents="agents"
    :tags="tags"
    :isLoading="isLoading"
    @model-toggle="handleModelToggle"
    @field-toggle="handleFieldToggle"
  />
</template>

<script setup lang="ts">
import ExportCheckboxs from '../ExportCheckboxs.vue';
import exportService from '@/services/api/resources/export/export';
import { useHumanResourceExport } from '@/store/modules/export/humanResource/export';
import { storeToRefs } from 'pinia';
import { onMounted, ref } from 'vue';

const humanResourceExport = useHumanResourceExport();
const { setModelFields, updateModelFieldSelection, toggleModelEnabled } =
  humanResourceExport;
const {
  model_fields,
  selected_fields,
  enabled_models,
  sectors,
  queues,
  agents,
  tags,
} = storeToRefs(humanResourceExport);
const isLoading = ref(false);

const fetchModelFields = async () => {
  isLoading.value = true;
  try {
    const modelFields = await exportService.getModelFields();
    const transformedFields: any = {};

    Object.entries(modelFields).forEach(([modelName, fields]) => {
      transformedFields[modelName] = {};
      Object.entries(fields as any).forEach(
        ([fieldName, fieldData]: [string, any]) => {
          transformedFields[modelName][fieldName] = {
            type: fieldData.type,
          };
        },
      );
    });

    setModelFields(transformedFields);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
};

const handleModelToggle = (modelName: string, enabled: boolean) => {
  toggleModelEnabled(modelName, enabled);
};

const handleFieldToggle = (
  modelName: string,
  fieldName: string,
  selected: boolean,
) => {
  updateModelFieldSelection(modelName, fieldName, selected);
};

onMounted(() => {
  fetchModelFields();
});
</script>
