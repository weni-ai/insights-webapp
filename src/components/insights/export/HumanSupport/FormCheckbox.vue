<template>
  <ExportCheckboxs
    data-testid="human-support-form-checkbox"
    :modelFields="model_fields"
    :selectedFields="selected_fields"
    :enabledModels="enabled_models"
    :modelFilters="modelFilters"
    :isLoading="isLoading"
    @model-toggle="handleModelToggle"
    @field-toggle="handleFieldToggle"
  />
</template>

<script setup lang="ts">
import ExportCheckboxs from '../ExportCheckboxs.vue';
import exportService from '@/services/api/resources/export/humanSupport/export';
import { useHumanSupportExport } from '@/store/modules/export/humanSupport/export';
import { storeToRefs } from 'pinia';
import { onMounted, ref, computed } from 'vue';

const humanSupportExport = useHumanSupportExport();
const { setModelFields, updateModelFieldSelection, toggleModelEnabled } =
  humanSupportExport;
const {
  model_fields,
  selected_fields,
  enabled_models,
  sectors,
  queues,
  agents,
  tags,
} = storeToRefs(humanSupportExport);
const isLoading = ref(false);

const modelFilters = computed(() => [
  {
    modelName: 'sectors',
    filterData: sectors.value,
  },
  {
    modelName: 'queues',
    filterData: queues.value,
  },
  {
    modelName: 'users',
    filterData: agents.value,
  },
  {
    modelName: 'sector_tags',
    filterData: tags.value,
  },
]);

const fetchModelFields = async () => {
  isLoading.value = true;
  try {
    const modelFields = await exportService.getModelFields();
    const transformedFields: any = {};

    Object.entries(modelFields).forEach(([modelName, fields]) => {
      transformedFields[modelName] = {};
      Object.entries(fields as any).forEach(
        ([fieldName, fieldData]: [string, any]) => {
          transformedFields[modelName][fieldName] = {
            type: fieldData.type,
          };
        },
      );
    });

    setModelFields(transformedFields);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
};

const handleModelToggle = (modelName: string, enabled: boolean) => {
  toggleModelEnabled(modelName, enabled);
};

const handleFieldToggle = (
  modelName: string,
  fieldName: string,
  selected: boolean,
) => {
  updateModelFieldSelection(modelName, fieldName, selected);
};

onMounted(() => {
  fetchModelFields();
});
</script>
