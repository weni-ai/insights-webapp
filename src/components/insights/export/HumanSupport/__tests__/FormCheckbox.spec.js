import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import FormCheckbox from '../FormCheckbox.vue';

const mockStore = {
  model_fields: { value: {} },
  selected_fields: { value: {} },
  enabled_models: { value: [] },
  sectors: { value: [] },
  queues: { value: [] },
  agents: { value: [] },
  tags: { value: [] },
  setModelFields: vi.fn(),
  updateModelFieldSelection: vi.fn(),
  toggleModelEnabled: vi.fn(),
};

vi.mock('@/store/modules/export/humanSupport/export', () => ({
  useHumanSupportExport: () => mockStore,
}));

vi.mock('@/services/api/resources/export/humanSupport/export', () => ({
  default: {
    getModelFields: vi.fn().mockResolvedValue({
      rooms: { field1: { type: 'string' } },
      users: { field2: { type: 'number' } },
    }),
  },
}));

vi.mock('pinia', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    storeToRefs: (store) => store,
  };
});

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: { en: {} },
});

config.global.plugins = [i18n];

describe('HumanSupport FormCheckbox', () => {
  let wrapper;

  const createWrapper = (storeOverrides = {}) => {
    Object.assign(mockStore, storeOverrides);

    return mount(FormCheckbox, {
      global: {
        stubs: {
          ExportCheckboxs: true,
        },
      },
    });
  };

  beforeEach(() => {
    vi.clearAllMocks();

    Object.assign(mockStore, {
      model_fields: { value: {} },
      selected_fields: { value: {} },
      enabled_models: { value: [] },
      sectors: { value: [] },
      queues: { value: [] },
      agents: { value: [] },
      tags: { value: [] },
    });

    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render ExportCheckboxs component', () => {
      const component = wrapper.find(
        '[data-testid="human-support-form-checkbox"]',
      );
      expect(component.exists()).toBe(true);
    });
  });

  describe('Lifecycle hooks', () => {
    it('should fetch model fields on mounted', async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
      expect(mockStore.setModelFields).toHaveBeenCalled();
    });

    it('should handle fetch errors', async () => {
      const consoleError = vi.spyOn(console, 'error').mockImplementation();
      const exportService = await import(
        '@/services/api/resources/export/humanSupport/export'
      );
      exportService.default.getModelFields.mockRejectedValueOnce(
        new Error('Fetch error'),
      );

      wrapper = createWrapper();
      await new Promise((resolve) => setTimeout(resolve, 0));

      expect(consoleError).toHaveBeenCalled();
      consoleError.mockRestore();
    });
  });

  describe('Computed properties', () => {
    it('should compute modelFilters with correct structure', () => {
      wrapper = createWrapper({
        sectors: { value: [{ value: '1', label: 'Sector 1' }] },
        queues: { value: [{ value: '2', label: 'Queue 1' }] },
        agents: { value: [{ value: '3', label: 'Agent 1' }] },
        tags: { value: [{ value: '4', label: 'Tag 1' }] },
      });

      const filters = wrapper.vm.modelFilters;
      expect(filters).toHaveLength(4);
      expect(filters[0].modelName).toBe('sectors');
      expect(filters[1].modelName).toBe('queues');
      expect(filters[2].modelName).toBe('users');
      expect(filters[3].modelName).toBe('sector_tags');
    });

    it('should use empty arrays when no filters', () => {
      const filters = wrapper.vm.modelFilters;
      expect(filters[0].filterData).toEqual([]);
    });
  });

  describe('Event handlers', () => {
    it('should handle model toggle', () => {
      wrapper.vm.handleModelToggle('rooms', true);
      expect(mockStore.toggleModelEnabled).toHaveBeenCalledWith('rooms', true);
    });

    it('should handle field toggle', () => {
      wrapper.vm.handleFieldToggle('rooms', 'field1', true);
      expect(mockStore.updateModelFieldSelection).toHaveBeenCalledWith(
        'rooms',
        'field1',
        true,
      );
    });
  });

  describe('Loading state', () => {
    it('should start with isLoading false', () => {
      expect(wrapper.vm.isLoading).toBe(false);
    });
  });

  describe('Component structure', () => {
    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
