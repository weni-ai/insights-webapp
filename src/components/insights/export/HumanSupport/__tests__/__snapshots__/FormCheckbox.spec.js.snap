// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HumanSupport FormCheckbox > Component structure > should match snapshot 1`] = `
<export-checkboxs-stub
  data-testid="human-support-form-checkbox"
  enabledmodels="[object Object]"
  isloading="false"
  modelfields="[object Object]"
  modelfilters="[object Object],[object Object],[object Object],[object Object]"
  selectedfields="[object Object]"
/>
`;
